### ISSUE TITLE
Null object exception

### ISSUE THREAD

**Issue by HurryStarfish** [GWRon/SimpleCodeTester#4]
I downloaded SCT and tried to use it on the tests that come with NG.

I changed the bottom part of sct.bmx to
```
'adjust compiler path for this test class
TTestCompiler.baseConfig.Add("bmk_path", "C:\Programmierung\BlitzMax-NG\bin\bmk.exe")

'adjust base config for all instances of that type
TTestCompiler.baseConfig.Add("app_type", "console")
TTestCompiler.baseConfig.Add("app_arch", "x86") 'unused yet
TTestCompiler.baseConfig.Add("debug", "0")
TTestCompiler.baseConfig.Add("threaded", "0")
TTestCompiler.baseConfig.Add("deleteBinaries", "1") 'delete binaries afterwards
TTestCompiler.baseConfig.Add("make_mods", "0") 
TTestCompiler.baseConfig.Add("quick", "0") 
TTestCompiler.baseConfig.fileUri = "baseConfig"

Global testManager:TTestManager = New TTestManager
testManager.AddTestsFromDirectory("C:\Programmierung\BlitzMax-NG\src\bcc\tests")
testManager.RunTests()
```
and ran it, but it caused a null object exception in function StandardIOAvailable. Same thing happens if I use the bmk path to the legacy BlitzMax compiler instead (I compiled SCT in legacy as well).
Am I doing something wrong here?

**Comment by GWRon** [comment]
Will try this evening. What OS do you run? Am not sure if windows10 limits accesses between "different" basepaths.

**Comment by HurryStarfish** [comment]
I use Windows 7.
Access permissions shouldn't be an issue. Plus, it's running several tests successfully before crashing:

> Building sct
> Compiling:sct.bmx
> flat assembler  version 1.68  (1048575 kilobytes memory)
> 3 passes, 3460 bytes.
> Linking:sct.debug.exe
> Executing:sct.debug.exe
> === STARTING TESTS ===
> * AMOUNT OF TESTS: 54
> * RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/array_01.bmx
>   COMMAND: C:\Programmierung\BlitzMax\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/array_01.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/array_01.bmx"
>   -> OK
> 
> * RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/array_slice.bmx
>   COMMAND: C:\Programmierung\BlitzMax\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/array_slice.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/array_slice.bmx"
>   -> OK
> 
> * RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/arrays_0.bmx
>   COMMAND: C:\Programmierung\BlitzMax\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/arrays_0.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/arrays_0.bmx"
>   -> OK
> 
> * RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/arrays_02.bmx
>   COMMAND: C:\Programmierung\BlitzMax\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/arrays_02.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/arrays_02.bmx"
>   -> OK
> 
> * RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/arrays_03.bmx
>   COMMAND: C:\Programmierung\BlitzMax\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/arrays_03.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/arrays_03.bmx"
>   -> OK
> 
> * RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/comparison_01.bmx
>   COMMAND: C:\Programmierung\BlitzMax\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/comparison_01.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/comparison_01.bmx"
>   -> OK
> 
> * RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/dotdot_01.bmx
>   COMMAND: C:\Programmierung\BlitzMax\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/dotdot_01.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/dotdot_01.bmx"
>   -> OK
> 
> * RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/end_01.bmx
>   COMMAND: C:\Programmierung\BlitzMax\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/end_01.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/end_01.bmx"
>   -> OK
> 
> * RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/exception_01.bmx
>   COMMAND: C:\Programmierung\BlitzMax\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/exception_01.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/exception_01.bmx"
>   -> OK
> 
> * RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/fp_arrays_01.bmx
>   COMMAND: C:\Programmierung\BlitzMax\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/fp_arrays_01.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/fp_arrays_01.bmx"
> 
> Process terminated
>

**Comment by GWRon** [comment]
You could run sct in maxide to debug it there...maybe it exposes what is failing.

But as said I will try on my own when home. Did never really try it on windows.

**Comment by GWRon** [comment]
I just tried it - and in the "ifthen_01.bmx" sample/test it took a long while (so 10 seconds or so) and ended with a segfault (when run as debug from within MaxIDE with the "Program - command line" menu option configured accordingly to run the tests).

I then executed it via `gdb` and it finished the tests:

```
=== FINISHED TESTS ===
* TIME: 30664ms
* FAILED: 1
* ERROR: 6
* OK: 53
[Inferior 1 (process 9500) exited normally]
``` 

Cleared the test folder, run it again from within MaxIDE ... and same results as when run from within gdb, so no "segfault".

Will try tomorrow on another machine. If you can "repeatingly" receive this test aborts, open up sct.bmx in MaxIDE, enable "debug", and then adjust "program - command line" to contain the path to the tests. "base.conf" within the test directory could be used to adjust "relative URI to bmk" from your SCT folder (so might be something like "../../../BlitzMax-NG/bin/bmk.exe")

**Comment by HurryStarfish** [comment]
Running sct from MaxIDE, with sct.bmx modified like in my initial post, I consistently get it when its trying to test `fp_arrays_01.bmx`.
The code in question is
```
	Method StandardIOAvailable:Int()
		Return standardIO.bufferpos Or standardIO.readAvail()
	End Method
```
with the callstack `RunTests` -> `Run` -> `Validate` -> `Eof` -> `IOAvailable` -> `StandardIOAvailable`.
`standardIO` is Null, hence the exception.

**Comment by GWRon** [comment]
While we could fix that crash by checking for "standardio and..." before doing the other stuff, it does not fix the underlying problem.

Something makes the stream not useable/openable.

**Comment by GWRon** [comment]
Hmm, the standardIO variable is created within `TCodeTesterProcess.RunProcess()`.

```BlitzMax
	Method RunProcess:TCodeTesterProcess()
		Local infd:Int, outfd:Int, errfd:Int

		handle = fdProcess(name, Varptr infd, Varptr outfd, Varptr errfd, flags)
		If Not handle Then Return Null

'		standardIO = TPipeStreamUTF8.Create(infd, outfd)
'		errorIO = TPipeStreamUTF8.Create(errfd, outfd)
		standardIO = TPipeStream.Create(infd, outfd)
		errorIO = TPipeStream.Create(errfd, outfd)

		processes.AddLast(Self)
		Return Self
	End Method
```

Maybe we run into some kind of "spawn/fork" limit or so?

Aside of that I am not sure if we properly finish created processes (maybe some of them "lock up") and so things might lead to a "concurrent spawned processes"-limit or so? Am currently checking that

**Comment by GWRon** [comment]
I committed some "cleanup + throw in case of null" lines - but this won't fix it for now.

I use "vanilla" to compile in non-threaded mode. With "NG" you automatically build the threaded variant. The threaded builds for now have issues - which I cannot explain yet.

The issue is, that the streams reading standardIO and errorIO receive data from the other tests.

```
* RUNNING TEST: ../../../Tools/BlitzMaxNG/bin/tests/framework/language/end_01.bmx
  COMMAND: /home/ronny/Arbeit/Programmieren/Tools/BlitzMaxNG/bin/bmk makeapp -a -t console -r -g x86 -o "../../../Tools/BlitzMaxNG/bin/tests/framework/language/end_01" "../../../Tools/BlitzMaxNG/bin/tests/framework/language/end_01.bmx"
  -> OK
[ 10%] Processing:exception_01.bmx
[ 90%] Compiling:exception_01.bmx.console.release.linux.x86.c
* RUNNING TEST: ../../../Tools/BlitzMaxNG/bin/tests/framework/language/fp_arrays_01.bmx
  COMMAND: /home/ronny/Arbeit/Programmieren/Tools/BlitzMaxNG/bin/bmk makeapp -a -t console -r -g x86 -o "../../../Tools/BlitzMaxNG/bin/tests/framework/language/fp_arrays_01" "../../../Tools/BlitzMaxNG/bin/tests/framework/language/fp_arrays_01.bmx"
  VALIDATION: FAILED
  expected:
  -----
  func1 1 1 1.0000000000000000
  func2 2 2 2.0000000000000000
  func3 3 3 3.0000000000000000
  ------
  received:
  -----
  ------
  -> FAILED
* RUNNING TEST: ../../../Tools/BlitzMaxNG/bin/tests/framework/language/function_01.bmx
  COMMAND: /home/ronny/Arbeit/Programmieren/Tools/BlitzMaxNG/bin/bmk makeapp -a -t console -r -g x86 -o "../../../Tools/BlitzMaxNG/bin/tests/framework/language/function_01" "../../../Tools/BlitzMaxNG/bin/tests/framework/language/function_01.bmx"
  -> OK
[100%] Linking:exception_01
* RUNNING TEST: ../../../Tools/BlitzMaxNG/bin/tests/framework/language/fp_local_01.bmx
  COMMAND: /home/ronny/Arbeit/Programmieren/Tools/BlitzMaxNG/bin/bmk makeapp -a -t console -r -g x86 -o "../../../Tools/BlitzMaxNG/bin/tests/framework/language/fp_local_01" "../../../Tools/BlitzMaxNG/bin/tests/framework/language/fp_local_01.bmx"
  -> OK
* RUNNING TEST: ../../../Tools/BlitzMaxNG/bin/tests/framework/language/exception_01.bmx
  COMMAND: /home/ronny/Arbeit/Programmieren/Tools/BlitzMaxNG/bin/bmk makeapp -a -t console -r -g x86 -o "../../../Tools/BlitzMaxNG/bin/tests/framework/language/exception_01" "../../../Tools/BlitzMaxNG/bin/tests/framework/language/exception_01.bmx"
  VALIDATION: FAILED
  expected:
  -----
  start TRY
  assignment to null is not allowed
  end TRY
  ------
  received:
  -----
  ------
  -> FAILED
```
You see multiple issues there: the received strings are most often "empty" instead of containing valid data.
Also it receives output of other threads when compiling a validation-binary. It should not do this:
```
* RUNNING TEST: ../../../Tools/BlitzMaxNG/bin/tests/framework/language/end_01.bmx
  COMMAND: /home/ronny/Arbeit/Programmieren/Tools/BlitzMaxNG/bin/bmk makeapp -a -t console -r -g x86 -o "../../../Tools/BlitzMaxNG/bin/tests/framework/language/end_01" "../../../Tools/BlitzMaxNG/bin/tests/framework/language/end_01.bmx"
  -> OK
[ 10%] Processing:exception_01.bmx
[ 90%] Compiling:exception_01.bmx.console.release.linux.x86.c
* RUNNING TEST: ../../../Tools/BlitzMaxNG/bin/tests/framework/language/fp_arrays_01.bmx
```
(as you see it talks about "exception_01.bmx" which another test. Another test then contains the [100%] of this exception_01 compilation).

I am not sure but I assume that "pub.mod/FreeProcess.mod" might not be threadsafe (PIDs of the executed commands) or something with its TPipeStream (on which the PipeStream in SCT is based on).

Nonetheless I think "threads" and your "null stream" might be two different bugs.

**Comment by HurryStarfish** [comment]
> I use "vanilla" to compile in non-threaded mode.
I've tried it with legacy/vanilla in both threaded and non-threaded, as well as in debug and release mode. The error happens regardless.

**Comment by HurryStarfish** [comment]
> Hmm, the standardIO variable is created within TCodeTesterProcess.RunProcess().
> 
> Maybe we run into some kind of "spawn/fork" limit or so?
> 
> Aside of that I am not sure if we properly finish created processes (maybe some of them "lock up") and so things might lead to a "concurrent spawned processes"-limit or so? Am currently checking that

That might be the case. When I run these tests, there are a lot of calls to TCodeTesterProcess.RunProcess happening (one or two for each test), but none to TCodeTesterProcess.Close. So pipe streams are continuously being created without being cleaned up up until the crash.

**Comment by GWRon** [comment]
If a process is ending thr handle gets nulled...and the process is removed from a container list (tlist) but maybe GC is not kicking in fast enough to get rid of the streams.

The new version should close streams earlier (on forked process end). Does it help?

**Comment by HurryStarfish** [comment]
I don't see a Finalizer in TCodeTesterProcess or in TPipeStream itself to clean up the unmanaged handles it allocates, so I don't think the GC could do anything here.

I just tried the new commit, and it ran to completion (with lots of failed tests and a crash in one of the tests but not from SCT itself).
However, it somehow managed to kill my explorer.exe - or rather, made one explorer window freeze and several others glitch out so badly that I had to kill it myself. Tried to take a screenshot of it, but that glitched as well and didnt even show the broken windows. That's quite impressive.
Now that I'm taking a close look at PUB.FreeProcess, that thing looks like a ticking timebomb that is badly in need of an update for NG. It's using `int`s everywhere to pass around handles (which are pointers), so it's not too surprising that things might go wrong with that. Though I can't know for sure if that's actually the reason for what I just experienced, of course...

**Comment by HurryStarfish** [comment]
I don't see a Finalizer in TCodeTesterProcess or in TPipeStream itself to clean up the unmanaged handles it allocates, so I don't think the GC could do anything here.

I just tried the new commit, and it ran to completion (with lots of failed tests and a crash in one of the tests but not from SCT itself).
However, it somehow managed to kill my explorer.exe - or rather, made one explorer window freeze and several others glitch out so badly that I had to kill it myself. Tried to take a screenshot of it, but that glitched as well and didnt even show the broken windows. That's quite impressive.
Now that I'm taking a closer look at PUB.FreeProcess, that thing seems like it is badly in need of an update for NG. It's using `ints` everywhere to pass around handles (which are pointers), so it's not too surprising that things might go wrong with that. That doesn't seem to explain what I just experienced however, since I just realized I was using legacy BlitzMax for this test run...

**Comment by GWRon** [comment]
Streams should get closed once the process is no longer "alive". There is some function called to remove "zombies". This also calls CloseStreams() if the process ended meanwhile ... which removes connections to the streams.

Maybe you could add another "print" to the functions outputting the "processes.Count()" and also a debugprint when streams get closed.

Maybe on your computer this does not happen at all?

**Comment by GWRon** [comment]
If you open up "base.processes.bmx" and replace

```BlitzMax
	Method CloseStreams()
```
with
```BlitzMax
	Method CloseStreams()
		print "CloseStreams() - name: " + name
```

You _should_ see this message printed during execution:
```
./bmk makeapp -t console -quick -r -x "/home/ronny/Arbeit/Programmieren/Projekte/BlitzMaxNG.project/SimpleCodeTester-master/sct.bmx"
[ 51%] Processing:base.processes.bmx
[..]
[100%] Linking:sct
Executing:sct
=== STARTING TESTS ===
* AMOUNT OF TESTS: 65
* RUNNING TEST: ../../../Tools/BlitzMaxNG/bin/tests/framework/language/array_01.bmx
  COMMAND: /home/ronny/Arbeit/Programmieren/Tools/BlitzMaxNG/bin/bmk makeapp -a -t console -r -g x86 -o "../../../Tools/BlitzMaxNG/bin/tests/framework/language/array_01" "../../../Tools/BlitzMaxNG/bin/tests/framework/language/array_01.bmx"
CloseStreams() - name: /home/ronny/Arbeit/Programmieren/Tools/BlitzMaxNG/bin/bmk makeapp -a -t console -r -g x86 -o "../../../Tools/BlitzMaxNG/bin/tests/framework/language/array_01" "../../../Tools/BlitzMaxNG/bin/tests/framework/language/array_01.bmx"
  -> OK
* RUNNING TEST: ../../../Tools/BlitzMaxNG/bin/tests/framework/language/array_slice.bmx
[...]
```

If "CloseStreams()" is not called, then `fdProcessStatus` is still returning a value > 0 (values interpreted as "true" in BlitzMax).



You could also replace (in the same file):
```BlitzMax
	Method RunProcess:TCodeTesterProcess()
```
with
```BlitzMax
	Method RunProcess:TCodeTesterProcess()
		print "RunProcess: "+processes.Count()+" processes still running."
```
And it should always print:
```
RunProcess: 0 processes still running.
```
(at least it does so on my computer).

**Comment by GWRon** [comment]
This is what I experienced already when trying to write a simple IDE/Editor some years ago: on some computers the output of the processes was "truncated". I needed to add artificial "delays" to properly fetch all stuff. So sometimes the process was no longer existing ("ended") - but there was still output coming in - as if there was some kind of "buffer".

Dunno if that happens to you too: when running SCT on windows, I see a lot of "command prompts" (or "DOS-Windows" - "Dos-Fenster") for "gcc.exe" popping up. It looks like a window is opened, closed and another one (on the next spot "bottom right" of the previous one) is opened.
This then happens 10 times or so - until a test is "run".


Nonetheless "stdOut" is not threadsafe, so mixed output of threads is another issue to solve/tinker with.

**Comment by GWRon** [comment]
in "base.testbase.bmx" there is a `Method Run:int()`. Maybe you could try to replace the while-loop 

```BlitzMax
		While process.Alive()
			If process.IOAvailable()
[...]
			EndIf
		Wend
```
with
```BlitzMax
		While process.IOAvailable() or process.Alive()
[...]
		Wend
```
`Alive()` is already cleaning up once a process is finished (I took that from freeprocess.mod - while I would prefer to do an `if not process.Alive() then process.Close()` (or similar). Nonetheless above's change should go into the loop as soon as there is something to print - and only check for "alive or not" if there is nothing more to print. 
If there is some kind of "buffered output" (so also "delayed" a bit) this wont help.

**Comment by HurryStarfish** [comment]
I am getting spammed with command prompts, yes. The above change to Method Run doesn't seem to make any difference.

**Comment by GWRon** [comment]
But it should be safe to replace (in the same function)
```BlitzMax
		'run cleanup - eg removing temporary files
		CleanUp()
```
with
```Blitzmax
		'run cleanup - eg removing temporary files
		CleanUp()

		'if not done yet, finish up
		process.Close()
```
So it should close up things, if something odd is making the whole thing think it is no longer alive (without having closed stuff properly). As it is no time critical command it should not have bad effects to close it even if there is nothing more to close).


@ spammed prompts
So how could that be? We spawn a maximum of 2 processes:
- compilation and execution process (to check if compilation fails or not - aka "syntax/compiler checks"
- execute created binary (if there is a validation to do, means not in all cases)

We only call `fdProcess` if `RunProcess` is executed - and as you appended the processes.Count()-print, you would see that line spammed too, if `RunProcess` was somehow triggered a bunch of files.

As you are an experienced programmer you might tinker a bit with that process stuff too: create a bunch of processes (one by one) and see if it spams command prompts too.

**Comment by GWRon** [comment]
Hmmm can BMK.exe be the culprit? I mean, we call something which calls something (bmk calls GCC, links stuff and the likes). Maybe things bork up there?

**Comment by HurryStarfish** [comment]
Oh; no, I'm not saying I'm getting tons of command prompts per test. It does look like one or two per test.
Running the tests on the legacy/vanilla bmk produces truncated output as well, by the way.

**Comment by GWRon** [comment]
You should not get any new prompts at all...at least it is not intented to do so ;-)


And yes, the truncation in my IDE tries were also done with vanilla (NG wasnt born that time).


Will try to checkout some more things tomorrow.

**Comment by GWRon** [comment]
Think I found the reason for the truncated output.

Currently it does a "if alive() then process output" thing - with "alive()" only checking if the process is still doing something. But output processing is only done if something was alive in that moment. Means it might stop fetching still present output if the process just "stopped right before".  

Will change the corresponding code.

**Comment by GWRon** [comment]
I got it pretty much working - except for some compiler errors which make the "TProcess" (of freeprocess.mod)  stumble over some odd characters in the "output buffer".

I read from a "TPipeStream" as long as "readAvail()" returns > 0 or "bufferPos" is > 0 (which means it already fetched the data but it was not processed via "ReadLine()" or similar). Somehow this compiler errors lead to the buffer being filled with characters, but "ReadLine()" returning empty strings...

So many hours for investigating this stuff - redoing it here and there... hmpf.

**Comment by GWRon** [comment]
So why is this? Because "ReadLine()" does only return a string if it finds "chr(10)" (aka "~n"/newline).
Of course I could just read in the existing buffer - but how to know if there is still some other output coming in?

**Comment by GWRon** [comment]
Ok, I pushed a new commit.

Could you please try? On my virtual machine the tests on Windows take way way longer than on my linux machine. Maybe processes are not started/finished that quick in this VM - or on Windows - or the compilation process is taking a break meanwhile :).
 
My normal linux box needs about 30 seconds for all tests. My VM .. I aborted after 5 (at about 90% of the tests).


But interesting to see how different OSes define things like "NAN" or scientific numbers:
Linux/Mac:
Cos(90)  = 6.1232339957367660e-17
ASin(45) = nan

Windows:
Cos(90)  = 2.6535848171582721e-017
ASin(45) = -1.#IND000000000000
(not only a different number but "e-17" vs "e-017")

**Comment by HurryStarfish** [comment]
The test run takes about 25-30 seconds for me in total.
The problem appears to be fixed! I'm still getting a good number of failing and error-ing tests, but now it seems to be the tests who are at fault (some are just missing `Framework`s, some failing for different reasons) and SCT is doing its thing correctly. No new command prompts popping up anymore either. Nice :)

Time to write some Try-Catch-Finally tests.

**Comment by GWRon** [comment]
I added a new commit - side by side validation results (I got pretty much annoyed by the array_slice-test ;-).

If you desire to only test your tests - make another directory :-).

```
=== FINISHED TESTS ===
* TIME: 31541ms
* FAILED: 2
* ERROR: 6
* OK: 57
Kompilierung erfolgreich beendet.
```
error = compilation issues
failed = validation failed
ok = :+1:

**Comment by HurryStarfish** [comment]
> test files for things "differing" on the OSes

That was already possible by wrapping a test in `? Linux` or similar I guess; but this makes it easier and more efficient.
I was about to ask about the array slice test actually - since I've just noticed that this one is still behaving strangely. Theoretically, it should print one empty line after every "]" (and nowhere else), and it does so when I run it from MaxIDE. But the result SCT just gave me is this:
```
* RUNNING TEST: C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/array_slice.bmx
  COMMAND: C:\Programmierung\BlitzMax-NG\bin\bmk.exe makeapp -a -t console -r -g x86 -o "C:/Programmierung/BlitzMax-NG/src/bcc/tests/framework/language/array_slice.exe" "C:\Programmierung\BlitzMax-NG\src\bcc\tests/framework/language/array_slice.bmx"
  VALIDATION: FAILED
  OK |Before:                |Before:
  OK |[                      |[
  OK |  AAAAAAAA   :   8     |  AAAAAAAA   :   8
  OK |  BBBBBBBB   :   8     |  BBBBBBBB   :   8
  OK |  CCCCCCCC   :   8     |  CCCCCCCC   :   8
  OK |  DDDDDDDD   :   8     |  DDDDDDDD   :   8
  OK |  EEEEEEEE   :   8     |  EEEEEEEE   :   8
  OK |  FFFFFFFF   :   8     |  FFFFFFFF   :   8
  OK |  GGGGGGGG   :   8     |  GGGGGGGG   :   8
  OK |  HHHHHHHH   :   8     |  HHHHHHHH   :   8
  OK |]                      |]
  XX |                       |0 : 
  XX |0 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |]
  XX |]                      |
  XX |                       |1 : 
  XX |1 :                    |[
  XX |[                      |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |]
  XX |]                      |
  XX |                       |2 : 
  XX |2 :                    |[
  XX |[                      |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |]
  XX |]                      |
  XX |                       |3 : 
  XX |3 :                    |[
  XX |[                      |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |]
  XX |]                      |
  XX |                       |4 : 
  XX |4 :                    |[
  XX |[                      |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |]
  XX |]                      |
  XX |                       |5 : 
  XX |5 :                    |[
  XX |[                      |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |]
  XX |]                      |
  XX |                       |6 : 
  XX |6 :                    |[
  XX |[                      |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |]
  XX |]                      |
  XX |                       |7 : 
  XX |7 :                    |[
  XX |[                      |  HHHHHHHH   :   8
  XX |  HHHHHHHH   :   8     |]
  XX |]                      |
  XX |                       |0 : 
  XX |0 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |]
  XX |]                      |
  XX |                       |1 : 
  XX |1 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |]
  XX |]                      |
  XX |                       |2 : 
  XX |2 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |]
  XX |]                      |
  XX |                       |3 : 
  XX |3 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |]
  XX |]                      |
  XX |                       |4 : 
  XX |4 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |]
  XX |]                      |
  XX |                       |5 : 
  XX |5 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |]
  XX |]                      |
  XX |                       |6 : 
  XX |6 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |
  OK |]                      |]
  XX |                       |7 : 
  XX |7 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |  HHHHHHHH   :   8
  XX |  HHHHHHHH   :   8     |
  OK |]                      |]
  XX |                       |0 : 
  XX |0 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |
  OK |  HHHHHHHH   :   8     |  HHHHHHHH   :   8
  OK |]                      |]
  XX |                       |1 : 
  XX |1 :                    |[
  XX |[                      |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |  HHHHHHHH   :   8
  XX |  HHHHHHHH   :   8     |]
  XX |]                      |
  XX |                       |2 : 
  XX |2 :                    |[
  XX |[                      |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |  HHHHHHHH   :   8
  XX |  HHHHHHHH   :   8     |]
  XX |]                      |
  XX |                       |3 : 
  XX |3 :                    |[
  XX |[                      |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |
  OK |  FFFFFFFF   :   8     |  FFFFFFFF   :   8
  OK |  GGGGGGGG   :   8     |  GGGGGGGG   :   8
  OK |  HHHHHHHH   :   8     |  HHHHHHHH   :   8
  OK |]                      |]
  XX |                       |4 : 
  XX |4 :                    |[
  XX |[                      |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |  HHHHHHHH   :   8
  XX |  HHHHHHHH   :   8     |]
  XX |]                      |
  XX |                       |5 : 
  XX |5 :                    |[
  XX |[                      |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |  HHHHHHHH   :   8
  XX |  HHHHHHHH   :   8     |]
  XX |]                      |
  XX |                       |6 : 
  XX |6 :                    |[
  XX |[                      |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |  HHHHHHHH   :   8
  XX |  HHHHHHHH   :   8     |]
  XX |]                      |
  XX |                       |7 : 
  XX |7 :                    |[
  XX |[                      |  HHHHHHHH   :   8
  XX |  HHHHHHHH   :   8     |]
  XX |]                      |
  XX |                       |0 : 
  XX |0 :                    |[
  XX |[                      |  AAAAAAAA   :   8
  XX |  AAAAAAAA   :   8     |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |]
  XX |]                      |
  XX |                       |1 : 
  XX |1 :                    |[
  XX |[                      |  BBBBBBBB   :   8
  XX |  BBBBBBBB   :   8     |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |]
  XX |]                      |
  XX |                       |2 : 
  XX |2 :                    |[
  XX |[                      |  CCCCCCCC   :   8
  XX |  CCCCCCCC   :   8     |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |]
  XX |]                      |
  XX |                       |3 : 
  XX |3 :                    |[
  XX |[                      |  DDDDDDDD   :   8
  XX |  DDDDDDDD   :   8     |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |]
  XX |]                      |
  XX |                       |4 : 
  XX |4 :                    |[
  XX |[                      |  EEEEEEEE   :   8
  XX |  EEEEEEEE   :   8     |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |]
  XX |]                      |
  XX |                       |5 : 
  XX |5 :                    |[
  XX |[                      |  FFFFFFFF   :   8
  XX |  FFFFFFFF   :   8     |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |]
  XX |]                      |
  XX |                       |6 : 
  XX |6 :                    |[
  XX |[                      |  GGGGGGGG   :   8
  XX |  GGGGGGGG   :   8     |  HHHHHHHH   :   8
  XX |  HHHHHHHH   :   8     |]
  XX |]                      |
  OK |                       |
  
  -> FAILED
```

**Comment by GWRon** [comment]
I am at exactly the same test case. I assume it is the "readline()" and some skipping of empty lines or so.

**Comment by GWRon** [comment]
Open up base.processes.bmx and replace ReadStandardIO with this

```BlitzMax
	Method ReadStandardIO:String()
		If StandardIOAvailable()
			local res:string = standardIO.ReadLine().Replace("~r","") '.Replace("~n","")
			'if res.length = 0 then return ReadStreamBuffer(standardIO)
			return res
		EndIf
	End Method
```
I did it the same way as the "errorIO" one - as BMK outputs some error lines with no "~n" at the end. Which means you cannot "Readline()" there. Will check this evening if I really need to do an additional "read" after the process-while-loops.

**Comment by GWRon** [comment]
As said I will look at it again this evening ... so it properly reads "empty lines".

**Comment by GWRon** [comment]
Pushed some more commits, should work now - reading validation-binary-execution output now as "blob" instead of "per line", needed to remove the last character (somehow on my linux and windows I get another "~n" which is not printed by Blitzmax but the OS or so).

Somehow the whole thing does not look as clean as it should (reading lines, reading from buffers, ...). Hope it works.

**Comment by HurryStarfish** [comment]
I just got a null object exception from [this line](https://github.com/GWRon/SimpleCodeTester/blob/17028462a6d80ec921db23da642408edeaf5412e/app.testcompiler.bmx#L163) - binaryProcess.standardIO was null.
